src/TrackGenerator.cpp

```diff
@@ -1554,6 +1557,64 @@ void TrackGenerator::correctFSRVolume(int fsr_id, FP_PRECISION fsr_volume) {
 
 
 /**
+ * @brief Generates the numerical centroids of the FSRs.
+ * @details This routine generates the numerical centroids of the FSRs
+ *          by weighting the average x and y values of each segment in the
+ *          FSR by the segment's length and azimuthal weight. The numerical
+ *          centroid fomula can be found in R. Ferrer et. al. "Linear Source
+ *          Approximation in CASMO 5", PHYSOR 2012.
+ */
+void TrackGenerator::generateFSRCentroids(){
+
+  int num_FSRs = _geometry->getNumFSRs();
+
+  /* Get FSR Volumes */
+  FP_PRECISION* FSR_volumes = getFSRVolumes();
+
+  /* Create array of centroids and initialize to origin */
+  Point** centroids = new Point*[num_FSRs];
+  for (int r=0; r < num_FSRs; r++){
+    centroids[r] = new Point();
+    centroids[r]->setCoords(0.0, 0.0);
+  }
+
+  /* Generate the fsr centroids */
+  for (int i=0; i < _num_azim; i++) {
+    for (int j=0; j < _num_tracks[i]; j++) {
+
+      int num_segments = _tracks[i][j].getNumSegments();
+      segment* segments = _tracks[i][j].getSegments();
+      double x = _tracks[i][j].getStart()->getX();
+      double y = _tracks[i][j].getStart()->getY();
+      double phi = _tracks[i][j].getPhi();
+
+      for (int s=0; s < num_segments; s++) {
+        segment* curr_segment = &segments[s];
+        int fsr = curr_segment->_region_id;
+        double volume = FSR_volumes[fsr];
+        centroids[fsr]->setX(centroids[fsr]->getX() + _azim_weights[i] *
+                             (x + cos(phi) * curr_segment->_length / 2.0) *
+                             curr_segment->_length / FSR_volumes[fsr]);
+        centroids[fsr]->setY(centroids[fsr]->getY() + _azim_weights[i] *
+                             (y + sin(phi) * curr_segment->_length / 2.0) *
+                             curr_segment->_length / FSR_volumes[fsr]);
+
+        x += cos(phi) * curr_segment->_length;
+        y += sin(phi) * curr_segment->_length;
+      }
```

Comment:
The `x` and `y` variables are initialized to the beginning of the track or equivalently the beginning of the first segment. For each segment, we need to know the segment midpoint, so when looping over the segments the midpoint is found by taking the segment starting point (`x` and `y`) and adding half the segment length times the appropriate sin/cos of the azimuthal angle. `x` and `y` are then incremented at the end of the loop so they become the starting point of the next segment. After the centroid is incremented for the last segment, `x` and `y` are set to the endpoint of the last segment. 

In short, I believe it is correct as is.